tests/components/statistics/test_sensor.py

```diff
@@ -1701,3 +1702,60 @@ async def test_device_id(
     statistics_entity = entity_registry.async_get("sensor.statistics")
     assert statistics_entity is not None
     assert statistics_entity.device_id == source_entity.device_id
+
+
+async def test_update_before_load(recorder_mock: Recorder, hass: HomeAssistant) -> None:
+    """Verify that updates happening before reloading from the database are handled correctly."""
+
+    current_time = dt_util.utcnow()
+
+    # enable and pre-fill the recorder
+    await hass.async_block_till_done()
+    await async_wait_recording_done(hass)
+
+    # enable and pre-fill the recorder
+    await hass.async_block_till_done()
+    await async_wait_recording_done(hass)
+
+    with (
+        freeze_time(current_time) as freezer,
+    ):
+        for value in VALUES_NUMERIC_LINEAR:
+            hass.states.async_set(
+                "sensor.test_monitored",
+                str(value),
+                {ATTR_UNIT_OF_MEASUREMENT: UnitOfTemperature.CELSIUS},
+            )
+            await hass.async_block_till_done()
+            current_time += timedelta(seconds=1)
+            freezer.move_to(current_time)
+
+        await async_wait_recording_done(hass)
+        # create the statistics component, get filled from database
+        assert await async_setup_component(
+            hass,
+            "sensor",
+            {
+                "sensor": [
+                    {
+                        "platform": "statistics",
+                        "name": "test",
+                        "entity_id": "sensor.test_monitored",
+                        "state_characteristic": "average_step",
```

Comment:
Hey @unfug-at-github could it be that this test case wrongly succeeds because you are using `average_step`? Will it start to fail if you switch to `mean`?

Reply:
It will fail if you switch to mean without changing the value as well. The mean of [1 .. 9] is 5 (= (1+2+3+4+5+6+7+8+9)/9), whereas average_step computes 4.5 (= (1+2+3+4+5+6+7+8)/8). With average_step the last value is not affecting the computation.

The important part is in both cases the value 10 that is added before the database finishes loading doesn't make it into the computation (otherwise the mean would be 5.5, and average_step 5).